### PR TITLE
Adds tests and ResponseHeader cleanup

### PR DESCRIPTION
**Pull request by jkotalik**
Exposing some light on some of the tests that will be added.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs [hunk 1]
```diff
@@ -60,12 +61,13 @@ internal unsafe HttpProtocol(PipeFactory pipeFactory, IntPtr pHttpContext)
             {
                 Method = GetVerb();
 
-                var version = GetVersion();
 
                 RawTarget = GetRawUrl();
-
-                HttpVersion = version.ToString();
-                Scheme = SslStatus.ToString();
+                // TODO version is slow.
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs]: Yes, we should be using cached strings

File: src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs [hunk 2]
```diff
@@ -96,6 +98,7 @@ internal unsafe HttpProtocol(PipeFactory pipeFactory, IntPtr pHttpContext)
 
                 // TODO: Avoid using long.ToString, it's pretty slow
                 RequestConnectionId = ConnectionId.ToString(CultureInfo.InvariantCulture);
+                UConnectionId = ConnectionId;
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs]: What is UConnectionId?

File: src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs [hunk 3]
```diff
@@ -110,8 +113,16 @@ internal unsafe HttpProtocol(PipeFactory pipeFactory, IntPtr pHttpContext)
                 // TODO: Parse socket ADDR for local and remote end point
                 // var localAddress = GetSocketAddress(pHttpRequest->Request.Address.pLocalAddress);
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs]: Remove the comments

File: src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs [hunk 4]
```diff
@@ -110,8 +113,16 @@ internal unsafe HttpProtocol(PipeFactory pipeFactory, IntPtr pHttpContext)
...
+                var localEndPoint = GetLocalEndPoint();
+                LocalIpAddress = localEndPoint.GetIPAddress();
+                LocalPort = localEndPoint.GetPort();
+
+                var remoteEndPoint = GetRemoteEndPoint();
+                RemoteIpAddress = remoteEndPoint.GetIPAddress();
+                LocalPort = remoteEndPoint.GetPort();
 
                 RequestHeaders = new RequestHeaders(this);
+                ResponseHeaders = new HeaderDictionary();
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs]: Put a TODO to optimize known response headers.

File: src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs [hunk 5]
```diff
@@ -122,7 +133,7 @@ internal unsafe HttpProtocol(PipeFactory pipeFactory, IntPtr pHttpContext)
             Output = new OutputProducer(pipe);
 
             // TODO: Only upgradable on Win8 or higher
-            _upgradeAvailable = true; // TODO
+            _upgradeAvailable = Environment.OSVersion.Platform == PlatformID.Win32NT && Environment.Version.Major >= 8;
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs]: Is this what we do in WebListener?

File: src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs [hunk 6]
```diff
@@ -147,13 +158,14 @@ internal unsafe HttpProtocol(PipeFactory pipeFactory, IntPtr pHttpContext)
...
-
+        public bool IsHttps { get; set; }
         public IPipe Input { get; set; }
         public OutputProducer Output { get; set; }
 
         public IHeaderDictionary RequestHeaders { get; set; }
-        public IHeaderDictionary ResponseHeaders { get; set; } = new HeaderDictionary();
+        public IHeaderDictionary ResponseHeaders { get; set; }
         internal HttpApiTypes.HTTP_VERB KnownMethod { get; }
+        internal ulong UConnectionId { get; }
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs]: Why do we need this?

File: src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs [hunk 7]
```diff
@@ -147,13 +158,14 @@ internal unsafe HttpProtocol(PipeFactory pipeFactory, IntPtr pHttpContext)
 
         public Stream RequestBody { get; set; }
         public Stream ResponseBody { get; set; }
-
+        public bool IsHttps { get; set; }
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs]: Why is this a property?

File: src/Microsoft.AspNetCore.Server.IIS/Microsoft.AspNetCore.Server.IIS.csproj [hunk 8]
```diff
@@ -21,4 +21,7 @@
     <PackageReference Include="Microsoft.AspNetCore.HttpSys.Sources" PrivateAssets="All" />
     <PackageReference Include="System.Security.Principal.Windows" />
   </ItemGroup>
+  <ItemGroup>
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IIS/Microsoft.AspNetCore.Server.IIS.csproj]: Remove this

File: src/Microsoft.AspNetCore.Server.IIS/Properties/AssemblyInfo.cs [hunk 9]
```diff
@@ -0,0 +1,3 @@
+// Copyright (c) .NET Foundation. All rights reserved.
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IIS/Properties/AssemblyInfo.cs]: Why is this file empty?

File: test/IISIntegration.FunctionalTests/IISIntegration.FunctionalTests.csproj [hunk 10]
```diff
@@ -20,4 +20,8 @@
     <PackageReference Include="xunit.runner.visualstudio" />
   </ItemGroup>
 
+  <ItemGroup>
```
Comment (davidfowl) [inline test/IISIntegration.FunctionalTests/IISIntegration.FunctionalTests.csproj]: Remove this?

File: src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs (hunk 2 above)
Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs]: It was used by HttpSys for RequestAborted, but I'll probably take the logic from Kestrel here.

File: test/IISIntegration.IISServerFunctionalTests/Http.config [hunk 11]
```diff
@@ -253,7 +253,7 @@
       <add name="ManagedEngine64" image="%windir%\Microsoft.NET\Framework64\v2.0.50727\webengine.dll" preCondition="integratedMode,runtimeVersionv2.0,bitness64" />
       <add name="ManagedEngineV4.0_32bit" image="%windir%\Microsoft.NET\Framework\v4.0.30319\webengine4.dll" preCondition="integratedMode,runtimeVersionv4.0,bitness32" />
       <add name="ManagedEngineV4.0_64bit" image="%windir%\Microsoft.NET\Framework64\v4.0.30319\webengine4.dll" preCondition="integratedMode,runtimeVersionv4.0,bitness64" />
-      <add name="AspNetCoreModule" image="C:\\Users\\jukotali\\code\\aspnetcoremodule\\artifacts\\build\\AspNetCore\\bin\\Release\\x64\aspnetcore.dll" />
+      <add name="AspNetCoreModule" image="%IIS_BIN%\aspnetcore.dll"/>
```
Comment (davidfowl) [inline test/IISIntegration.IISServerFunctionalTests/Http.config]: 😄

File: src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs (hunk 5 above)
Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs]: Probably not, Is there one?

File: test/IISTestSite/StartupFeatureCollection.cs [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+                        Assert.True(IPAddress.IsLoopback(context.Connection.RemoteIpAddress));
+                        context.Connection.RemoteIpAddress = IPAddress.IPv6Any;
+                        Assert.Equal(IPAddress.IPv6Any, context.Connection.RemoteIpAddress);
+                        await context.Response.WriteAsync("Success");
+                        return;
+                    }
+                }
+                catch (Exception exception)
+                {
+                    await context.Response.WriteAsync(exception.ToString());
```
Comment (davidfowl) [inline test/IISTestSite/StartupFeatureCollection.cs]: Return 500?

File: test/IISTestSite/StartupLargeResponseBody.cs [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+    public class StartupLargeResponseBody
+    {
+        public void Configure(IApplicationBuilder app, ILoggerFactory loggerFactory)
+        {
+            app.Run(async context =>
+            {
+                if (context.Request.Path.Equals("/LargeResponseBody"))
+                {
+                    // TODO is there a better way to remove the ? from the query string
+                    if (int.TryParse(context.Request.QueryString.Value.Substring(1), out var length))
```
Comment (davidfowl) [inline test/IISTestSite/StartupLargeResponseBody.cs]: What is this trying to do?

Comment (jkotalik) [reply test/IISTestSite/StartupLargeResponseBody.cs]: Send a request with the length of the responsebody.

Comment (davidfowl) [reply test/IISTestSite/StartupLargeResponseBody.cs]: Why not just read the specific query string key then?

File: src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs (hunk 2 above)
Comment (davidfowl) [reply src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs]: No need to copy from kestrel, I'm just trying to understand why we have 2 properties:
 UConnectionId and ConnectionId.

Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs]: I'll drop it for now until I implement the feature.

File: test/IISIntegration.IISServerFunctionalTests/ShutdownTests.cs [hunk 14]
```diff
@@ -0,0 +1,9 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Server.IIS.FunctionalTests
+{
+    class ShutdownTests
```
Comment (davidfowl) [inline test/IISIntegration.IISServerFunctionalTests/ShutdownTests.cs]: Empty stub?

File: src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs [hunk 15]
```diff
@@ -313,13 +343,15 @@ public unsafe void CreateResponseHeader(bool appCompleted)
 
             HttpApiTypes.HTTP_RESPONSE_V2* pHttpResponse = NativeMethods.http_get_raw_response(_pHttpContext);
 
-            _pinnedHeaders = SetHttpResponseHeaders(pHttpResponse);
+            // We should be sending the headers here as there are many responses that don't have status codes.
+            var pinnedHeaders = SerializeHeaders(pHttpResponse);
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs]: Uh what is this?

File: src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs [hunk 16]
```diff
@@ -543,6 +615,7 @@ public void StartWritingResponseBody()
 
         private unsafe IISAwaitable WriteAsync(ReadableBuffer buffer)
         {
+            // TODO I am super uncertain about the correctness of this method.
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs]: Not sure this comment is helpful.

File: src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs (hunk 15 above)
Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs]: Sorry, was working on another feature and decided to wait for another PR.

File: test/IISIntegration.IISServerFunctionalTests/Http.config (hunk 11 above)
Comment (Tratcher) [reply test/IISIntegration.IISServerFunctionalTests/Http.config]: Not using the nuget deployed one yet?

File: src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs [hunk 17]
```diff
@@ -147,14 +167,27 @@ internal unsafe HttpProtocol(PipeFactory pipeFactory, IntPtr pHttpContext)
...
-        public IHeaderDictionary ResponseHeaders { get; set; } = new HeaderDictionary();
+        public IHeaderDictionary ResponseHeaders { get; set; }
         internal HttpApiTypes.HTTP_VERB KnownMethod { get; }
 
+        public int StatusCode
+        {
+            get { return _statusCode; }
+            set
+            {
+                // Http.Sys automatically sends 100 Continue responses when you read from the request body.
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs]: this comment seems out of place. Did there use to be a check that prohibited 100 status codes and lower?

File: src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs [hunk 18]
```diff
@@ -121,20 +133,28 @@ internal unsafe HttpProtocol(PipeFactory pipeFactory, IntPtr pHttpContext)
             var pipe = _pipeFactory.Create(new PipeOptions { ReaderScheduler = TaskRunScheduler.Default });
             Output = new OutputProducer(pipe);
 
-            // TODO: Only upgradable on Win8 or higher
-            _upgradeAvailable = true; // TODO
+            // TODO: Also make this not slow
+            _upgradeAvailable = (Environment.OSVersion.Version >= _win8Version);
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs]: _upgradeAvailable can be static, you don't need to calculate it per request.

File: src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs [hunk 19]
```diff
@@ -107,11 +113,17 @@ internal unsafe HttpProtocol(PipeFactory pipeFactory, IntPtr pHttpContext)
...
-                // TODO: Parse socket ADDR for local and remote end point
-                // var localAddress = GetSocketAddress(pHttpRequest->Request.Address.pLocalAddress);
-                // var remoteAddress = GetSocketAddress(pHttpRequest->Request.Address.pRemoteAddress);
+                var localEndPoint = GetLocalEndPoint();
+                LocalIpAddress = localEndPoint.GetIPAddress();
+                LocalPort = localEndPoint.GetPort();
+
+                var remoteEndPoint = GetRemoteEndPoint();
+                RemoteIpAddress = remoteEndPoint.GetIPAddress();
+                LocalPort = remoteEndPoint.GetPort();
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs]: remote port

File: src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs (hunk 1 above)
Comment (Tratcher) [reply src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs]: Take the code from HttpSys

**Comment by jkotalik** [comment]
@Tratcher @davidfowl Once this is in I'm going to polish up a PR for merging with dev.

File: test/IISIntegration.IISServerFunctionalTests/Http.config (hunk 11 above)
Comment (jkotalik) [reply test/IISIntegration.IISServerFunctionalTests/Http.config]: I'll add that as a TODO for now.

**Comment by jkotalik** [comment]
Making PR with dev.

**Pull request closed without merging**